constantine-go/constantine.go

```diff
@@ -265,4 +266,377 @@ func (ctx EthKzgContext) VerifyBlobKzgProofBatchParallel(tp Threadpool, blobs []
+			C.GoString(C.ctt_codec_ecc_status_to_string(status)),
+		)
+		return false, err
+	}
+	return true, nil
+}
+
+func (pub *EthBlsPubKey) DeserializeCompressed(src [48]byte) (bool, error) {
+	status := C.ctt_eth_bls_deserialize_pubkey_compressed((*C.ctt_eth_bls_pubkey)(pub),
+		(*C.byte)(unsafe.Pointer(&src[0])),
+	)
+	if status != C.cttCodecEcc_Success && status != C.cttCodecEcc_PointAtInfinity {
+		err := errors.New(
+			C.GoString(C.ctt_codec_ecc_status_to_string(status)),
+		)
+		return false, err
+	}
+	return true, nil
+}
+
+func (sig *EthBlsSignature) DeserializeCompressed(src [96]byte) (bool, error) {
+	status := C.ctt_eth_bls_deserialize_signature_compressed((*C.ctt_eth_bls_signature)(sig),
+		(*C.byte)(unsafe.Pointer(&src[0])),
+	)
+	if status != C.cttCodecEcc_Success && status != C.cttCodecEcc_PointAtInfinity {
+		err := errors.New(
+			C.GoString(C.ctt_codec_ecc_status_to_string(status)),
+		)
+		return false, err
+	}
+	return true, nil
+}
+
+func (sig *EthBlsSignature) Sign(sec EthBlsSecKey, message []byte) (bool, error) {
+	C.ctt_eth_bls_sign((*C.ctt_eth_bls_signature)(sig), (*C.ctt_eth_bls_seckey)(&sec),
+		(*C.byte)(unsafe.Pointer(&message[0])),
+		(C.ptrdiff_t)(len(message)),
+	)
+	return true, nil
+}
+
+func FastAggregateVerify(pubkeys []EthBlsPubKey, message []byte, aggregate_sig EthBlsSignature) (bool, error) {
+	if len(pubkeys) == 0 {
+		err := errors.New(
+			"No public keys given.",
+		)
+		return false, err
+	}
+	status := C.ctt_eth_bls_fast_aggregate_verify((*C.ctt_eth_bls_pubkey)(unsafe.Pointer(&pubkeys[0])),
+		(C.ptrdiff_t)(len(pubkeys)),
+		(*C.byte)(unsafe.Pointer(&message[0])),
+		(C.ptrdiff_t)(len(message)),
+		(*C.ctt_eth_bls_signature)(&aggregate_sig),
+	)
+	if status != C.cttEthBls_Success {
+		err := errors.New(
+			C.GoString(C.ctt_eth_bls_status_to_string(status)),
+		)
+		return false, err
+	}
+	return true, nil
+}
+
+// Go wrapper of the `ctt_span` type of the C API
+type CttSpan struct {
+	data *C.byte
+	len  C.size_t
+}
+
+// NOTE: this does not work, because it would end up with us having a
+// Go pointer we pass to C, which points to Go pointers.
+func toSpan(data *[]byte) (span C.ctt_span) { //CttSpan) {
+	span = C.ctt_span{data: (*C.byte)(unsafe.Pointer(&(*data)[0])), len: C.size_t(len(*data))} // in place span
+	return span
+}
+
+func newSpans(data [][]byte) (spans []CttSpan) {
+	// TODO: this is ugly, but due to Go's rules about not allowing to
+	// pass a Go pointer, which itself contains a Go pointer, we cannot
+	// pass the data of each `message` to C... So gotta malloc and copy
+	spans = make([]CttSpan, len(data), len(data))
+	for i, msg := range data {
+		mem := C.malloc(C.ulong(len(msg)))
+		C.memcpy(mem, unsafe.Pointer(&msg[0]), C.ulong(len(msg)))
+		spans[i] = CttSpan{data: (*C.byte)(mem), len: C.size_t(len(msg))}
+	}
+	return spans
+}
+
+func freeSpans(spans []CttSpan){
+	// frees the malloc'd memory of the given spans
+	for _, msg := range spans {
+		C.free(unsafe.Pointer(msg.data))
+	}
+}
```

Comment:
If we're happy with the copy approach for now, I'd go ahead and remove the not working `toSpan` here.